### PR TITLE
Fix quota check for single shared files

### PR DESCRIPTION
**Pull request by PVince81**
- [x] TODO: web UI
- [x] TODO: fix quota plugin for Webdav
- [x] TODO: ~~make part file quota check on the correct (shared) storage instead of local one~~ disable quota check on part files
#### Single file local share
- [x] TEST: upload small file to owner storage with no space left on uploader's storage => 507
- [x] TEST: upload big file to owner storage with no space left on uploader's storage => 507
- [x] TEST: upload small file to overwrite received share file with no space left on uploader's storage => works
- [x] TEST: upload big file to overwrite received share file with no space left on uploader's storage => works
- [x] TEST: web UI overwrite file with no space left on uploader's storage
#### Single file federated share
- [x] TEST: upload small file to owner storage with no space left on uploader's storage => 507
- [x] TEST: upload big file to owner storage with no space left on uploader's storage => 507
- [x] TEST: upload small file to overwrite received share file with no space left on uploader's storage => works
- [x] TEST: upload big file to overwrite received share file with no space left on uploader's storage => works
- [x] TEST: web UI overwrite file with no space left on uploader's storage
#### Other
- [x] TEST: regression test quota in general
- [x] TEST: regression test quota with folder shares
- [x] TEST: regression test quota with fed folder shares

@icewind1991

### CODE REVIEW

**Comment by PVince81** [comment]
Fix for https://github.com/owncloud/core/issues/20686

**Comment by PVince81** [comment]
@icewind1991 I suggest reverting https://github.com/owncloud/core/pull/16159 and using a different approach. Goal is to have the part file on the same storage as the final file, to make sure the quota check applies to the correct storage. (in fed sharing cases no part file is used, so it's only for local file shares). Do you agree ?

**Comment by icewind1991** [comment]
> Goal is to have the part file on the same storage as the final file

That's not possible for single file storages (such as shares)

**Comment by PVince81** [comment]
> That's not possible for single file storages (such as shares)

Dammit, before lunch I couldn't see it and now that you mention it, I see why you're right.

Hmmm, then we need a different way for checking the quota. Currently it's the quota stream wrapper of the part storage that kicks in.

**Comment by PVince81** [comment]
Rebased and fixed the web UI: in the web UI we now defer the quota check when we detect that we're trying to overwrite an existing received share file. Ugly but works... In 9.1 with Webdav upload https://github.com/owncloud/core/pull/21237/files this will go away again as the quota check will be server-only.

**Comment by PVince81** [comment]
@icewind1991 should we then disable part files when overwriting a single file share ?
In the case of federated shares the remote server already uses its own part file.

The problem is mostly for local shares. No part file means we overwrite an existing file and lock it during the transfer instead of during the final rename. It also means that if the transfer fails for some reason, the original file is lost :frowning:

**Comment by PVince81** [comment]
I did a quick test locally and if I disable the part file for such files, it works.
However we'd need to add additional locking instructions to make sure that we use an exclusive lock directly on the final file during the transfer, for local shares.

In the case of fed shares, the lock would be done anyway on the remote server, so it wasn't needed so far.

File: apps/dav/lib/connector/sabre/file.php [hunk 1]
```diff
@@ -107,9 +107,17 @@ public function put($data) {
...
-		list($partStorage) = $this->fileView->resolvePath($this->path);
+		list($partStorage, $internalPartStoragePath) = $this->fileView->resolvePath($this->path);
 		$needsPartFile = $this->needsPartFile($partStorage) && (strlen($this->path) > 1);
 
+		if (!empty($this->path) && empty($internalPartStoragePath)) {
+			// this is a single shared file acting as a storage, disable part file because the quota
+			// check would happen on the wrong storage
+
+			// TODO: exclusive lock here and release at the end!
+			$needsPartFile = false;
```
Comment (PVince81) [inline apps/dav/lib/connector/sabre/file.php]: @icewind1991 what do you think ?

**Comment by PVince81** [comment]
Maybe there's another way, but it would be more complex: disable quota wrappers for single shared files, and add an explicit quota check inside the upload code. Would be ugly though.

**Comment by PVince81** [comment]
Talked with @schiesbn and @nickvergessen and we found a better idea: make the quota stream wrapper let part files go through (no quota check). Disabling part files isn't a good idea anyway.

Considering that we already have quota checks on the upload endpoints, it would work.
Apps using the node API wouldn't have part files anyway.

Let's have a try with this approach instead.

**Comment by PVince81** [comment]
It worked ! https://github.com/owncloud/core/commit/c18d56ced945b06d75ab84c622bc7f1d7eb0a132

@icewind1991 any objections to this approach ? See my previous comment.

**Comment by rullzer** [comment]
Aaah that looks a lot safer than disablig the `.part` file from single file shares!

**Comment by icewind1991** [comment]
I'm fine with this approach

**Comment by PVince81** [comment]
Okay thanks, let me squash this an check the unit tests.

**Comment by PVince81** [comment]
Cannot test fed sharing overwriting, there seems to be a bug on master.
Ticket incoming.

**Comment by PVince81** [comment]
- [ ] Blocked by https://github.com/owncloud/core/issues/22433

**Comment by PVince81** [comment]
- [ ] fix unit tests

**Comment by PVince81** [comment]
Fixed unit tests and also added the long-awaited chunking test, had to do some gymnastics to mock the OC_FileChunking class... https://github.com/owncloud/core/commit/05f866d358811a58c05e7af772614a16e145cd33#diff-e5121d9fe91be53ed25766c25eea058dR118

**Comment by PVince81** [comment]
Rebased onto master, this is now properly testable.

Please review @MorrisJobke @icewind1991 @rullzer @nickvergessen @schiesbn

File: apps/files/ajax/upload.php [hunk 2]
```diff
@@ -135,15 +135,13 @@
...
-}
-if ($maxUploadFileSize >= 0 and $totalSize > $maxUploadFileSize) {
-	OCP\JSON::error(array('data' => array('message' => $l->t('Not enough storage available'),
-		'uploadMaxFilesize' => $maxUploadFileSize,
-		'maxHumanFilesize' => $maxHumanFileSize)));
-	exit();
+$isReceivedShare = isset($_REQUEST['isReceivedShare']) && $_REQUEST['isReceivedShare'] === 'true';
+// defer quota check for received shares
+if (!$isReceivedShare) {
+	$totalSize = 0;
```
Comment (nickvergessen) [inline apps/files/ajax/upload.php]: this var is not defined in all cases anymore. should be outside of th eif

Comment (PVince81) [inline apps/files/ajax/upload.php]: right, I deleted too much

File: apps/files/ajax/upload.php [hunk 3]
```diff
@@ -135,15 +135,13 @@
...
-$totalSize = 0;
-foreach ($files['size'] as $size) {
-	$totalSize += $size;
-}
-if ($maxUploadFileSize >= 0 and $totalSize > $maxUploadFileSize) {
-	OCP\JSON::error(array('data' => array('message' => $l->t('Not enough storage available'),
-		'uploadMaxFilesize' => $maxUploadFileSize,
-		'maxHumanFilesize' => $maxHumanFileSize)));
-	exit();
+$isReceivedShare = isset($_REQUEST['isReceivedShare']) && $_REQUEST['isReceivedShare'] === 'true';
```
Comment (nickvergessen) [inline apps/files/ajax/upload.php]: We have \OC::$server->getRequest(), so we don't have to touch globals anymore :-O

Also I'm not sure what this means, anyone can set this on his request? But I think it just removes the warning and finishes uploading?

File: apps/files/ajax/upload.php (hunk 2 above)
Comment (nickvergessen) [inline apps/files/ajax/upload.php]: Btw I just saw, this variable is not used anymore after this?

Comment (PVince81) [inline apps/files/ajax/upload.php]: You're right. I deleted more than too much.
Was supposed to keep the error message from above.

File: apps/files/ajax/upload.php (hunk 3 above)
Comment (PVince81) [inline apps/files/ajax/upload.php]: If you skip the warning with this setting and it turns out there is not enough space after all, then the quota stream wrapper will catch it and you'll get a "could not upload" error. I didn't want to fix too much of ajax/upload.php as this file is going to be killed in 9.1 with webdav upload.

File: apps/files/ajax/upload.php [hunk 4]
```diff
@@ -136,8 +136,12 @@
 $maxHumanFileSize = OCP\Util::humanFileSize($maxUploadFileSize);
 
 $totalSize = 0;
-foreach ($files['size'] as $size) {
-	$totalSize += $size;
+$isReceivedShare = \OC::$server->getRequest()->getParam('isReceivedShare', false) === 'true';
```
Comment (PVince81) [inline apps/files/ajax/upload.php]: Fixed

File: apps/dav/lib/connector/sabre/quotaplugin.php [hunk 5]
```diff
@@ -111,6 +113,11 @@ public function checkQuota($uri, $data = null) {
 		return true;
 	}
 
+	public function getFileChunking($info) {
+		// FIXME: need a factory for better mocking support
```
Comment (nickvergessen) [inline apps/dav/lib/connector/sabre/quotaplugin.php]: You limbo-:dancer:

**Comment by nickvergessen** [comment]
Fixes the issue :+1: 
Oracle is unhappy just because it's no patch for itself.

**Comment by icewind1991** [comment]
:+1: looks good

**Comment by PVince81** [comment]
Regression: https://github.com/owncloud/core/issues/22899

**Comment by PVince81** [comment]
Looks like this indirectly messed up for other stuff.
Now for federated shares we also query the quota directly on the remote file, and the Webdav doesn't return quota values for files, which is expected.

In 9.0.2 I'll look into an alternative approach where we go back to check the quota on the parent folder, and only check it on a file if we know it's a shared file. (file mount points are so awful...)

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/23546 for a better fix

**Pull request merged**